feat: customization of exposed service ports

vsoch opened the pull request:
Problem: when deploying usernetes in a multi-tenant environment, there will likely be clashing of ports due to their hard coding of traditional values.
Solution: expose these ports (those mapped to the host in the docker-compose.yaml and kubeadm-config yaml as environment variables to customize.

@AkihiroSuda - I think the kubeadm-config.yaml should maintain the internal port 6443? But let me know if this needs to be added to the `$(NODE_SHELL)`. I tested this just now (with defaults) and it seemed to work, so minimally (superficially) it isn't breaking anything. Please let me know if there is somewhere else these should be provided.

This will close #345

AkihiroSuda commented on Makefile:
```diff
@@ -5,6 +5,12 @@ HOSTNAME ?= $(shell hostname)
 export HOSTNAME := $(HOSTNAME)
 
+# Exposed ports to the host (default to same as in container)
+export U7S_ETCD_EXPOSED_PORT ?= 2379
```
Probably these envs should rather have the common prefix (U7S_EXPOSED_PORT_ or perhaps just U7S_PORT_)

AkihiroSuda commented on docker-compose.yaml:
```diff
@@ -12,13 +12,13 @@ services:
     ports:
       # etcd
-      - 2379:2379
+      - 2379:${U7S_ETCD_EXPOSED_PORT}
```
How was this tested?
The ordering seems opposite.

vsoch replied:
You're right - they are flipped! I set them to the same thing so often I didn't realize it wasn't `<container>:<host>`

AkihiroSuda commented on docker-compose.yaml:
kubeadm-config has to be updated.
A port client doesn't magically know the server port number

AkihiroSuda reviewed (changes requested):
This doesn't seem to work.
How was this tested?

vsoch commented:
@AkihiroSuda it only worked when I didn't customize the kube-apiserver port - I haven't tested this setup that I just pushed.

vsoch commented:
It works in my setup:

![image](https://github.com/user-attachments/assets/bb395c38-fa61-4799-85f7-627e39bec6ac)


Can confirm the commit:

![image](https://github.com/user-attachments/assets/3c599648-89b2-4c59-ae4d-716b14f7d520)

So I think I need to reproduce the tests - there is probably something happening in CI that isn't happening in my setup.

vsoch commented:
ok, fixed. I had fat fingered an extra `:` with one of the docker-compose directives accidentally before commit! It wasn't in my testing environment because the setup had my fork of the repository just before that mistake was added. Sorry about that.

For the lima multi-node test, I noticed that 6443 is hard coded in a couple of places:

```console
.github/workflows/main.yaml:          ssh -q -f -N -L 6443:127.0.0.1:6443 -F ~/.lima/host0/ssh.config lima-host0
hack/create-cluster-lima.sh:ssh -q -f -N -L 6443:127.0.0.1:6443 -F ~/.lima/host0/ssh.config lima-host0
```
A note for the future - if we extend testing to customizing the ports, we'd need the CI to change these ports too.

vsoch commented:
I chose to keep the prefix `U7S_EXPOSED_PORT_*` instead of `U7S_PORT_*` anticipating some future where customization might be desired for the internal set. The first is longer, but it's more obvious to the developer user which port (of the set) is being referred to.

AkihiroSuda commented:
>  we extend testing to customizing the ports

Yes, this should be tested on the CI with multiple nodes

vsoch commented:
> Yes, this should be tested on the CI with multiple nodes

Gotcha. I'll start working on this, and tell you when it's ready for review (I'll likely push a few broken ones to test how it works and don't want to waste your time reviewing something that isn't ready).

vsoch commented:
@AkihiroSuda it looks like lima is just running the same logic, but in lima, so arguably it's already tested that the default is not broken. Since it is currently using the default (6443) if I change to a custom port, then we lose the default test in multi-node. So do you want the lima test to customize the port (unlike the single container) and lose the testing of the current default, or do you want another test added just for testing the port customization?

AkihiroSuda commented:
Should be another test

vsoch commented:
Not working yet - this is what I'm debugging (the difference between working and not working setups):

```
[kubeconfig] Using kubeconfig folder "/etc/kubernetes"
W1205 03:57:04.933499     144 endpoint.go:57] [endpoint] WARNING: port specified in controlPlaneEndpoint overrides bindPort in the controlplane address
[kubeconfig] Writing "admin.conf" kubeconfig file
W1205 03:57:05.214923     144 endpoint.go:57] [endpoint] WARNING: port specified in controlPlaneEndpoint overrides bindPort in the controlplane address
[kubeconfig] Writing "super-admin.conf" kubeconfig file
W1205 03:57:05.399882     144 endpoint.go:57] [endpoint] WARNING: port specified in controlPlaneEndpoint overrides bindPort in the controlplane address
[kubeconfig] Writing "kubelet.conf" kubeconfig file
W1205 03:57:05.551592     144 endpoint.go:57] [endpoint] WARNING: port specified in controlPlaneEndpoint overrides bindPort in the controlplane address
[kubeconfig] Writing "controller-manager.conf" kubeconfig file
W1205 03:57:05.755102     144 endpoint.go:57] [endpoint] WARNING: port specified in controlPlaneEndpoint overrides bindPort in the controlplane address
```
I know the controlPlaneAddres [is here](https://github.com/rootless-containers/usernetes/blob/31887a806a777aef492c9c94098a1e5379a11c25/kubeadm-config.yaml#L10) and I'm still looking for the other one - maybe it's something in the kind image?

vsoch commented:
@AkihiroSuda I need to ask for your expertise/eyes here - I've been trying to change the value and I keep finding other places where defaults are set, but now I'm out of ideas! Here is what I currently have:

https://github.com/rse-ops/usernetes/pull/1

I'll keep looking, and please let me know if you see something I'm missing!

vsoch commented:
I'm going to try the first idea again of just maintaining the internal port to be 6443 and see how far I get. I think I'll need to generate another kubeconfig, and will run into trouble if there is some kind of hashing of the port/address somewhere (and they are different).

vsoch commented:
okay got a little further that way, going to bed now (1am!) Here is the last run: https://github.com/rse-ops/usernetes/actions/runs/12175446833/job/33959176701

Basically, where it goes wrong is with the kubeadm-join I think? It might be a catch-22 - we need the join-command to have the exposed port (since that's what the control plane is showing) but then that port is also expected to be used internally for API information. At least when I tested changing the kubeadm-config.yaml right before it to use the exposed port instead of 6443, it showed that it was hitting the control plane, but then internally trying to use the wrong port. Let me know if you have other ideas - I need to go to sleep. Thanks for the help today!

vsoch commented:
My current attempt is trying to update the kubeadm-config.yaml as well, because (I think) there are extra params that will populate manifests under `/etc/kubernetes/manifests`. I am having even less luck with this approach - just about everything is failing. https://github.com/rse-ops/usernetes/pull/1/files

vsoch commented:
I'm going to take a break, but I'll be back for another attempt later today :)

vsoch commented:
@AkihiroSuda it's all working! To summarize the changes here and my work today:

### Strategy

I've been testing two strategies today:

- changing the external mapped port
- changing both ports

There are pros and cons to both I ultimately decided on the first, primarily because there are cases when something is generated inside of the container (where the port might be 6443) and others when that thing is used externally (e.g., the join-command generated on the inside would then need a sed to change the port). I think as long as we hit all the config files where the api server port can be defined, the second is the less error prone strategy.

### etcd, flannel, etc.

Given the amount of debugging it took to change just the kube api-server port, I changed these variables to be hard coded in the docker-compose / Makefile, so they aren't currently customizable, but it's setup in the future for them to be. Each will need careful testing, etc., and it would be dangerous to just allow changing them without that.

### kubeadm-config

I kept seeing many errors about the deprecated kubeadm-config, and I realized that a lot of the problems I was facing had to do with fields in manifests that weren't updated in some manifest in `/etc/kubernetes/manifests`. I didn't like the idea of having some unknown translation layer between the version 3 and what winds up there (there were fields I couldn't seem to change with v3) so I decided on a strategy to do the update.  I hope this is a direction that is OK to go in - the new format is fairly nice, and doesn't issue those warnings. 

### The Bugs! :lady_beetle: 

There were 5 bugs that only revealed when I looked at a combination of pod logs, systemctl logs, and /var/log logs:

- An illegal flag (that resulted from the config update)
- A flannel error (because the cidr block wasn't being passed)
- The controller manager was missing that the cloud provider should be external
- I had added verbosity to the kubeadm init command, and podman-compose was (for some reason) picking it up
- Filesystem permission issue

### Testing

> This is possibly the coolest bit! :sunglasses: 

In order for this to be possible, we need to test the default (6443) and then a custom port. The best way to do that isn't to copy paste entire code blocks, but to use GitHub reusable workflows. That allows us to define variables that get piped into different variants of a workflow matrix, and results in what we see here:

![image](https://github.com/user-attachments/assets/636a6ed2-d416-442d-bf26-f78eaff061bc)


I love this setup.

vsoch commented on Makefile:
```diff
@@ -140,8 +149,7 @@ sync-external-ip:
 
 .PHONY: kubeadm-join
 kubeadm-join:
-	$(NODE_SHELL) sh -euc "envsubst </usernetes/kubeadm-config.yaml >/tmp/kubeadm-config.yaml"
```
Two notes here:

1. I made the `join-command` executed with bash so it will work in a read only filesystem
2. I wasn't sure if we needed the first command if the node isn't a control plane, because it seemed to generate that kubeadm-config.yaml in /tmp but then not do anything with it. Let me know if we do and I'll add back the line (and I'm curious to know how it's used).

AkihiroSuda commented on .github/workflows/main.yaml:
```diff
@@ -65,83 +65,14 @@ jobs:
+    uses: ./.github/workflows/reusable-multi-node.yaml
+    with:
+      # Defaults to 6443
+      kube_apiserver_port: "8080"
```
This should change all the ports, not just apiserver

vsoch replied:
@AkihiroSuda each of these needs to be done separately and carefully, and I think in different PRs. Based on the challenges I hit with just api server, it would increase the scope here enormously, which already is more than I intended. We are going to need all of these services customized in the next fiscal year so you can count on me to come back and do the changes. My preference is to do one at a time and have clearly scoped PRs. If you want me to just add to the template here I could, but it would be misleading because the current setup does not support that customization.

vsoch replied:
This is now done, with the exception of flannel, which is based on vxlan / a Linux default. See my comment in https://github.com/rootless-containers/usernetes/issues/345#issuecomment-2525465642

vsoch replied:
Flannel is added.

AkihiroSuda commented on docker-compose.yaml:
```diff
@@ -19,6 +19,7 @@ services:
       # kubelet (default: 10250)
       - ${U7S_PORT_KUBELET}:${U7S_PORT_KUBELET}
       # flannel (default: 8472)
+      # This is a vxlan default, so we don't change it
```
Why can't change?

vsoch replied:
It requires reconfiguring the system which is out of scope for my expertise and what I intended to do for this PR. E.g., https://blog.svedr.in/posts/configuring-a-vxlan-interface-on-a-linux-machine/

If you want I can expose the variable for the expert user to change, but I can’t set up a testing environment for this.

AkihiroSuda replied:
https://github.com/flannel-io/flannel/blob/v0.26.1/Documentation/backends.md#vxlan

> * `Port` (number): UDP port to use for sending encapsulated packets. On Linux, defaults to kernel default, currently 8472, but on Windows, must be 4789.

vsoch replied:
So you are saying if I expose it and just add the value in the flannel yaml that is applied here:

```yaml
  net-conf.json: |
    {
      "Network": "10.244.0.0/16",
      "EnableNFTables": false,
      "Backend": {
        "Type": "vxlan"
      }
    }
```
it will just work? It doesn't require some change to the host system unrelated to kubernetes?

AkihiroSuda replied:
Maybe use helm?
- https://github.com/flannel-io/flannel/issues/1426#issuecomment-1460407817

AkihiroSuda replied:
Or just plain yq and jq

vsoch replied:
@AkihiroSuda the helm chart values.yaml does not expose what we need, so I clone the repository and customize it for here, and have at least a start:

https://github.com/rse-ops/usernetes/pull/5

I think the missing piece is the changes to (kernel modules or the host network?) e.g., commands here: https://github.com/nehalineogi/azure-cross-solution-network-architectures/blob/main/advanced-linux-networking/linux-vxlan.md or here: https://docs.kernel.org/networking/vxlan.html

Could you help me to understand the minimum that I need to make this change?

vsoch replied:
I think I figured this out.

AkihiroSuda commented on kubeadm-config.yaml:
```diff
@@ -1,31 +1,47 @@
-  extraArgs:
-    advertise-address: "${U7S_HOST_IP}"
-    cloud-provider: external
-    # Default: "Hostname,InternalDNS,InternalIP,ExternalDNS,ExternalIP"
```
This comment seems lost

vsoch replied:
Added back: https://github.com/rootless-containers/usernetes/pull/346/commits/1166cf2f8b034e3bff515a7a1b1abab1f3ac91c5

AkihiroSuda commented on Makefile:
```diff
@@ -5,6 +5,15 @@ HOSTNAME ?= $(shell hostname)
 export HOSTNAME := $(HOSTNAME)
 
+# Exposed ports to the host (default remains in container)
+# These each need to be separately tested before being enabled, so hard coded for now
```
This comment seems outdated, except for flannel

vsoch replied:
Fixed.

AkihiroSuda commented:
Please squash the commits

vsoch commented:
@AkihiroSuda could you please request a re-run of this single test? https://github.com/rootless-containers/usernetes/actions/runs/12220225586/job/34087751036?pr=346 I am not able to.

AkihiroSuda commented on Dockerfile:
```diff
@@ -9,14 +9,17 @@ RUN arch="$(uname -m | sed -e s/x86_64/amd64/ -e s/aarch64/arm64/)" && \
-  rm -f "${fname}"
+  rm -f "${fname}" && \
+  curl -fsSL -o get_helm.sh https://raw.githubusercontent.com/helm/helm/main/scripts/get-helm-3 && \
+  chmod 700 get_helm.sh && ./get_helm.sh
```
```suggestion
  chmod 700 get_helm.sh && ./get_helm.sh && rm -f get_helm.sh
```

AkihiroSuda commented on Makefile.d/install-flannel.sh:
```diff
@@ -0,0 +1,86 @@
+cd $flannel_root/chart
+
+# Write a new values.yaml that exposes what we need
+cat <<EOF > ./new-values.yaml
```
Can we just use `yq` to replace `flannel.backendPort`?

vsoch replied:
In the helm chart, we need to re-create the values.yaml because the [majority is commented out](https://github.com/flannel-io/flannel/blob/master/chart/kube-flannel/values.yaml#L27). And that's the only file that has that variable exposed to go into the config map. If we tried to edit the single install file, it would be a much more hacky operation as we would need to not just parse yaml (the seconds and values aren't there) but json objects within it. I think the approach here with helm that you suggested is the best strategy for now. We can engage further with the maintainers there to see if 1. they can vendor the helm chart to have values exposed we need (and then we can just use it) or 2. vendor the single yaml install to include more of the defaults (that are currently missing).

AkihiroSuda replied:
No need to parse JSON if you apply `yq` to `values.yaml`.

Or maybe just use a plain old `diff` to patch `values.yaml`?

vsoch replied:
I think I prefer the strategy that we currently have because it gives absolute transparency into the values.yaml that we are applying. I think what I'd like to do after the work here is follow up on the flannel repository and ask the developers to expose the values, and then we can just use theirs without our customization here. So we can consider this (ideally) a temporary fix. Is that OK?

AkihiroSuda commented on docker-compose.yaml:
```diff
@@ -11,14 +11,16 @@ services:
+      # kubelet (default: 10250)
+      - ${U7S_PORT_KUBELET}:${U7S_PORT_KUBELET}
+      # flannel (default: 8472)
+      # This is a vxlan default, so we don't change it
```
The comment is outdated

AkihiroSuda commented on Makefile.d/install-flannel.sh:
```diff
@@ -0,0 +1,86 @@
+
+mv ./new-values.yaml ./kube-flannel/values.yaml
+
+kubectl create namespace kube-flannel || WARNING "kube-flannel namespace is already created"
```
This warning may cause a confusion when kubectl failed due to another error

vsoch replied:
What if we do a kubectl get pods first, and then this line? The reason I added this was because in testing, I wanted to re-install flannel, and when the namespace was still there (without that catch) it would fail.

vsoch replied:
I added a command to get pods (in that namespace, which could be informative) that would catch an issue with running kubectl, if there is one. And I'm then falling back to the warning so the script doesn't fail on a reinstall (when the namespace will exist).

AkihiroSuda commented on Makefile.d/install-flannel.sh:
```diff
@@ -0,0 +1,86 @@
+  args:
+  - "--ip-masq"
+  - "--kube-subnet-mgr"
+  - "--etcd-endpoints=\"http://127.0.0.1:4001,http://127.0.0.1:${U7S_PORT_ETCD}\""
```
This is unlikely to work when the node is not a control plane

vsoch replied:
Are you sure? Wouldn't the lima tests fail if this were the case?

vsoch replied:
The reason I did this is because it is the default - what is currently generated when we don't specify anything.

vsoch replied:
I added `--etcd-endpoints=\"http://127.0.0.1:4001,https://${U7S_HOST_IP}:${U7S_PORT_ETCD},http://127.0.0.1:${U7S_PORT_ETCD}` to try, which keeps the defaults the version we have would generate, plus a variant that I think would be see-able if not running on the node.

AkihiroSuda replied:
The etcd config should be just removed.

etcd isn't used here, as kube-subnet-mgr is enabled
https://github.com/flannel-io/flannel/blob/v0.26.1/Documentation/configuration.md

vsoch replied:
okay will remove - but just to clarify, when we do that, it does default to the original value that I had set. Is that OK?

vsoch replied:
okay I think I chose a good middle strategy - I commented it out, and left a note to the user that it isn't used with the link (your comment above). This way, if the developer user decides to edit the configuration and runs into an issue with etcd (having changed the port) they will have awareness into the issue. A lot of the problems I ran into were just needing to spend time figuring out command line flags (and how to provide them) so this might save someone time in the future!

AkihiroSuda commented on Makefile.d/install-flannel.sh:
```diff
@@ -0,0 +1,86 @@
+  args:
+  - "--ip-masq"
+  - "--kube-subnet-mgr"
+  - "--etcd-endpoints=\"http://127.0.0.1:4001,http://127.0.0.1:${U7S_PORT_ETCD}\""
```
What is 4001?

vsoch replied:
It looks to be an older port: 
![image](https://github.com/user-attachments/assets/11a4301f-bcfb-4a4f-b74f-176a442a3439)

And the reason I did this is because when it's absent, that is what I see - here is our same deployment here with that line commented out. This is the flannel logs showing the default etcd endpoints. So I think it is built into our kind image, for whatever reason, that older port is being generated there:

![image](https://github.com/user-attachments/assets/2aeb6023-b907-42d7-aece-af0de92c153f)

Note `http://127.0.0.1:2379`. Should we include the newer port (2380) and then do 127.0.0.1 and the `U7S_NODE_IP` to cover all cases? Or would you like to try something else? I just brought up a cluster.

vsoch commented:
This failure https://github.com/rootless-containers/usernetes/actions/runs/12226896350/job/34102844371?pr=346 might be ephemeral - let's wait for the rest of the tests to finish to see. The one change I made was that list of etcd endpoints to flannel.

AkihiroSuda commented on Makefile.d/install-flannel.sh:
```diff
@@ -0,0 +1,89 @@
+# Run this first in case a failure with kubectl
+kubectl get pods -n kube-flannel
+# Fall back to warning so a re-install does not fail
+kubectl create namespace kube-flannel || WARNING "kube-flannel namespace is already created"
```
This might be more robust
```suggestion
kubectl create namespace kube-flannel || WARNING "kube-flannel namespace might have been already created"
```

vsoch replied:
Done!

AkihiroSuda reviewed (changes requested):
The etcd config is wrong and should be removed from the flannel script.

Other parts mostly LGTM, thanks

AkihiroSuda reviewed (approved):
Thanks, will merge and open a follow-up PR(s) for cleaning up
